### PR TITLE
Upgrade Elasticsearch to 8

### LINKED ISSUE #31256: Migrate elasticsearch client to 8.* (or latest 7.*)
### Body

The eleasticsearch version we use is old (from 2021). It is eleasticsearch 7.* and since we are using elasticsearch-dsl, we cannot used elasticsearch 8 currently (first version of that was also released in 2021). 

We are currently limited to < 7.15.0 additionally to being limited to 7, because 7.15.0 breaks our tests as we found out when we released some of the dependencies as part of https://github.com/apache/airflow/pull/30067

We should migrate to a newer version of elasticsearch. There are two options:

1) Fix the tests and migrate to lates 7.* version - and bump to latest released 7* by fixing our tests (possible for short term).
2) Migreate to 8 version (better long term)

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by Owen-CH-Leung**
Fixes #31256 

Upgrade the `elasticsearch` provider to use official `elasticsearch` package with version >8.

Remote logging with latest ES package should also work with this PR merged. To test locally, rebuild the breeze image with elasticsearch constraint set up `elasticsearch>8,<9`, set up elasticsearch 8 , kibana & filebeat in docker respectively, with the following changes to the `airflow.cfg`

```
[logging]
remote_logging = True

[elasticsearch]
json_format = True
host = http://host.docker.internal:9200
host_field = host.name
offset_field = log.offset
```

Note that with the latest es package, host field under `elasticsearch` has to be a full URL starting with `http://` or `https://`. If you just put `host.docker.internal:9200` that will not work

### CODE REVIEW

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml:2382]: Why this change?

**Comment by potiuk** [comment]
Looks good, but it needs a bit "surrounding changes" and "context".

I think we need to think a bit carefully about the change/migration (configs?)  -and soem changelog documentation should be addded explaining what's going on and explaining context.

The Elasticsearch config can. now be moved to provider - this is where it belongs. We should keep the pre-2.7 defaults in wiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning but the whole "elasticsearch" config should be moved to provider.yaml - same as celery, kubernetes and others.

File: airflow/config_templates/config.yml
Comment (Owen-CH-Leung) [reply airflow/config_templates/config.yml:2382]: In elasticsearch 7, `use_ssl` is an accepted parameter when constructing `ElasticSearch` client. See the following source code : 

https://github.com/elastic/elasticsearch-py/blob/7.14/elasticsearch/client/__init__.py#L113

However, in elasticsearch 8, it no longer accepts `use_ssl` parameter. See the following source code: 

https://github.com/elastic/elasticsearch-py/blob/8.9/elasticsearch/_sync/client/__init__.py#L129

Therefore to make the testsuite compile with the ES8 , I use `http_compress` as the argument (which is one of the accepted arguments for constructing ES client

**Comment by Owen-CH-Leung** [comment]
> Looks good, but it needs a bit "surrounding changes" and "context".
> 
> I think we need to think a bit carefully about the change/migration (configs?) -and soem changelog documentation should be addded explaining what's going on and explaining context.
> 
> The Elasticsearch config can. now be moved to provider - this is where it belongs. We should keep the pre-2.7 defaults in wiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning but the whole "elasticsearch" config should be moved to provider.yaml - same as celery, kubernetes and others.

Thanks. I'll some more contexts into changelog soon.

For moving elasticsearch config, do you mean the following lines should be moved to `provider.yaml` ?

https://github.com/apache/airflow/blob/main/airflow/config_templates/config.yml#L2396-L2410

**Comment by potiuk** [comment]
> For moving elasticsearch config, do you mean the following lines should be moved to provider.yaml ?

https://github.com/apache/airflow/blob/main/airflow/config_templates/config.yml#L2396-L2410

Yes. See for example: https://github.com/apache/airflow/pull/32777  - this is a new thing that has been added in Airflow 2.7 - providers can now contribute their own whole "sections" and they can be removed from generic "airflow" configuration. 

Also, you will need to keep old "pre-2.7" defaults in https://github.com/apache/airflow/blob/main/airflow/config_templates/pre_2_7_defaults.cfg for back-compatibility (this one provides the default values for old providers that have not yet moved their config to ). The PR for Hive above does not have it yet as I realised it only afterwards. As you can see elasticsearch_configs is there, but I think - if we decide to also move "elasticsearch" to the provider, we need to add "elasticearch" to the defaults as well before 2.7.0 is released. 

I will go ahead and add it now, otherwise some back-compatibility scenarios might not work (old provider assuming default value retrieved might stop working).

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (pankajkoti) [inline airflow/providers/elasticsearch/CHANGELOG.rst]:
```suggestion
6.0.0
```

Since we are suggesting a breaking change, we need to have a major bump

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py:105]: May I know what error do we see if the protocol is not included in the set value?

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py:295]: Why do we have here a  type ignore?

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py:304]: same question regarding type ignore

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py:311]: Cannot we not narrow down the exception we catch? Is the previous exception no longer present? If so, have they added any other similar class and can we use that?

Having such a broad level exception catch and not re-raising it might lead to some silent failures.

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (pankajkoti) [inline airflow/providers/elasticsearch/CHANGELOG.rst]:
```suggestion
  Upgrade to Elasticsearch 8. The ElasticsearchTaskHandler & ElasticsearchSQLHook will now use Elasticsearch 8 package.
```

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (pankajkoti) [inline airflow/providers/elasticsearch/CHANGELOG.rst]:
```suggestion
  This also means that the release drops support for Elasticsearch 7 and below and will no longer work with
  Elasticsearch database that is below version 8.
```

Comment (eladkal) [inline airflow/providers/elasticsearch/CHANGELOG.rst]: I am not sure if this PR is a breaking change?
Airflow wise what exactly gets broken here?
Updating upstream package version by itself is not a breaking change.

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/log/es_task_handler.py:105]: In ES 8, when constructing a new ElasticSearch client, a full URL including http / https is expected, and you will get a `ValueError` when calling the `__init__` like below: 

image

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/log/es_task_handler.py:295]: So - if we look at the official ES package, the `count` API from the `ElasticSearch` class actually didn't have the `body` as parameter. See this link: https://github.com/elastic/elasticsearch-py/blob/main/elasticsearch/_sync/client/__init__.py#L801

But the body parameter is still accepted because there's a decorator at the beginning, which modifies the function to accept `body` as the argument. See this few lines : https://github.com/elastic/elasticsearch-py/blob/main/elasticsearch/_sync/client/__init__.py#L798-L800

Therefore, without type ignore, the pre-commit job `mypy at provider` will actually get failed because it thinks that body is not an accepted parameter (which actually is)

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/CHANGELOG.rst]: Ok - will revise accordingly

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/CHANGELOG.rst]: Ok. Will revise accordingly

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/CHANGELOG.rst]: Sorry - what I'm trying to say is that this release will drop support for ElasticSearch database version 7 or below since `elasticsearch 8` will not work with DB version below 8 (i.e. airflow users should also upgrade their ES database to version 8 to ensure compatibility). I wrote this because I read the following doc and learn that `Elasticsearch language clients are only backwards compatible with default distributions and without guarantees made`

https://elasticsearch-py.readthedocs.io/en/stable/

Airflow-wise, nothing should really get broken

File: airflow/providers/elasticsearch/log/es_task_handler.py
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/log/es_task_handler.py:311]: Yes the exception `elasticsearch.exceptions.ElasticsearchException` is no longer present. Instead new class of exceptions are defined such as `UnsupportedProductError`, `NotFoundError` and so on. See this file: 

https://github.com/elastic/elasticsearch-py/blob/main/elasticsearch/exceptions.py

And I feel like all those errors can occur when calling the ES API. So maybe we should raise the exception after logging to the error log ?

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (potiuk) [reply airflow/providers/elasticsearch/CHANGELOG.rst]: Yeah. @eladkal is right - dependency change is not a reason for major bump - even if it is such a "big" dependency. This very clear in  https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api:

> What should I do if I update my own dependencies without changing the public API?

> That would be considered compatible since it does not affect the public API. Software that explicitly depends on the same dependencies as your package should have their own dependency specifications and the author will notice any conflicts. Determining whether the change is a patch level or minor level modification depends on whether you updated your dependencies in order to fix a bug or introduce new functionality. We would usually expect additional code for the latter instance, in which case it’s obviously a minor level increment.

File: airflow/providers/elasticsearch/provider.yaml
Comment (potiuk) [inline airflow/providers/elasticsearch/provider.yaml:75]: On top of also moving `[elasticsearch]` section (which I thin makes sense) it should be accompanied by adding the link to elastisearch provider configuration in the Airflow documentation: (following https://github.com/apache/airflow/pull/32777). 

When we have more of those, we might want to choose to do it automatically but for now we need to add it "manually"

**Comment by Owen-CH-Leung** [comment]
Thanks all. I've moved `elasticsearch` section to provider.yaml, revise the descriptions in CHANGELOG and also revised the docs. For now I keep the release version as `5.1.0` but feel free to let me know if we want to make it a major bump

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Looks cool from my side. Thanks @Owen-CH-Leung for being so responsive to our comments :)

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (eladkal) [inline airflow/providers/elasticsearch/CHANGELOG.rst]: Please remove above including the version.
It is not known yet which version is going to be next.. it could be 5.0.1, 5.1.0, 6.0.0
we know the release version only in run time.

The change we need here is just the note after the changelog title.. during release the release manager set it to the proper classification

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/elasticsearch/CHANGELOG.rst
Comment (Owen-CH-Leung) [reply airflow/providers/elasticsearch/CHANGELOG.rst]: Sure. Removed.

**Comment by Owen-CH-Leung** [comment]
I think @pankajkoti is also testing on ES remote logging with ES 8. Let's merge the PR when he comes back to us

**Review by eladkal — APPROVED** [review]

**Comment by pankajkoti** [comment]
I can confirm ES remote logging works fine for me from this branch and Elasticsearch 8.9.0 🎉 . Thanks for your patience @Owen-CH-Leung

**Comment by eladkal** [comment]
Cool!
Will merge when CI is green

**Comment by potiuk** [comment]
> I can confirm ES remote logging works fine for me from this branch and Elasticsearch 8.9.0 🎉 . Thanks for your patience @Owen-CH-Leung

Fantastic news !

**Comment by potiuk** [comment]
aaand merged.

**Pull request merged**
